docs/api/synchronization_library/barrier.md

```diff
@@ -23,26 +23,35 @@ cuda::std::barrier<> bb;
 cuda::barrier<cuda::thread_scope_block> c;
 ```
 
-The class template `barrier` may also be declared without initialization; a
+The class template `barrier` may also be declared without initialization in the `cuda::` namespace; a
 friend function `init` may be used to initialize the object.
 
 ```c++
 // Shared memory does not allow initialization.
 __shared__ cuda::barrier<cuda::thread_scope_block> b;
```

Comment:
I don't know of a specification of the exact semantics of `__shared__` in CUDA C++ and how they interact with default constructors, but I imagine they are very similar to obtaining a pointer to a raw allocation with kernel lifetime, e.g., 

```c++
cuda::barrier<cuda::thread_scope_block>* b = ...points to uninitialized storage...;
```

where `init(b, 1)` acts like `new(b) cuda::barrier<cuda::thread_scope_block>{1}`.

Placement-new guarantees that the object will be initialized in place. Why do we need `init` ? Couldn't we implement the constructor as follows:

```c++
cuda::barrier<cuda::thread_scope_block>::barrier(ptrdiff_t expected) {
    init(this, expected);
}
```

and just use "placement `new`" for initializing barriers?

Reply:
That is more or less what happens; but having non-experts write placement new expressions is probably not the greatest idea.